File: Código-Fonte/flutter/lib/models/activity.dart
import 'package:MegaObra/utils/json_serializable.dart';

class ActivityCreate implements JsonSerializable {
  final String description;
  final int location_id;
  final int professional_amount;
  final int laborer_amount;
  final int professional_minutes;
  final int laborer_minutes;
  final int average_labor_cost;
  final DateTime start_date;
  final DateTime deadline_date;
  final int? project_id;
  final List<int> chunks;
  final List<int> restriction;
  final List<int> employee;
  final List<int> materials;
  final List<double> materials_quantity;
  final bool? rain;

  const ActivityCreate({
    required this.description,
    required this.location_id,
    required this.professional_amount,
    required this.laborer_amount,
    required this.professional_minutes,
    required this.laborer_minutes,
    required this.average_labor_cost,
    required this.start_date,
    required this.deadline_date,
    required this.chunks,
    required this.restriction,
    required this.employee,
    required this.materials,
    required this.materials_quantity,
    this.project_id,
    this.rain,
  });

  factory ActivityCreate.fromJson(Map<String, dynamic> json) {
    return ActivityCreate(
      description: json['description'] as String,
      location_id: json['location_id'] as int,
      professional_amount: json['professional_amount'] as int,
      laborer_amount: json['laborer_amount'] as int,
      professional_minutes: json['professional_minutes'] as int,
      laborer_minutes: json['laborer_minutes'] as int,
      average_labor_cost: json['average_labor_cost'] as int,
      start_date: DateTime.parse(json['start_date'] as String),
      deadline_date: DateTime.parse(json['deadline_date'] as String),
      project_id: json['project_id'] != null ? json['project_id'] as int : null,
      chunks: (json['chunks'] as List<dynamic>).map((e) => e as int).toList(),
      restriction: (json['restriction'] as List<dynamic>).map((e) => e as int).toList(),
      employee: (json['employee'] as List<dynamic>).map((e) => e as int).toList(),
      materials: (json['materials'] as List<dynamic>).map((e) => e as int).toList(),
      materials_quantity: (json['materials_quantity'] as List<dynamic>).map((e) => (e as num).toDouble()).toList(),
      rain: json['rain'] != null ? json['rain'] as bool : null,
    );
  }

  @override
  Map<String, dynamic> toJson() {
    final data = {
      'description': description,
      'location_id': location_id,
      'professional_amount': professional_amount,
      'laborer_amount': laborer_amount,
      'professional_minutes': professional_minutes,
      'laborer_minutes': laborer_minutes,
      'average_labor_cost': average_labor_cost,
      'start_date': start_date.toIso8601String(),
      'deadline_date': deadline_date.toIso8601String(),
      'chunks': chunks,
      'restriction': restriction,
      'employee': employee,
      'materials': materials,
      'materials_quantity': materials_quantity,
      'project_id': project_id,
      'rain': rain,
    };

    return data;
  }
}

class ActivityBase implements JsonSerializable {
  final String description;
  final int location_id;
  final int professional_amount;
  final int laborer_amount;
  final int professional_minutes;
  final int laborer_minutes;
  final int average_labor_cost;
  final bool done;
  final int created_by;
  final DateTime start_date;
  final DateTime deadline_date;
  final DateTime creation_date;
  final DateTime? done_date;

  const ActivityBase({
    required this.description,
    required this.location_id,
    required this.professional_amount,
    required this.laborer_amount,
    required this.professional_minutes,
    required this.laborer_minutes,
    required this.average_labor_cost,
    required this.done,
    required this.created_by,
    required this.creation_date,
    required this.start_date,
    required this.deadline_date,
    this.done_date,
  });

  factory ActivityBase.fromJson(Map<String, dynamic> json) {
    return ActivityBase(
      description: json['description'] as String,
      location_id: json['location_id'] as int,
      professional_amount: json['professional_amount'] as int,
      laborer_amount: json['laborer_amount'] as int,
      professional_minutes: json['professional_minutes'] as int,
      laborer_minutes: json['laborer_minutes'] as int,
      average_labor_cost: json['average_labor_cost'] as int,
      done: json['done'] as bool,
      created_by: json['created_by'] as int,
      start_date: DateTime.parse(json['start_date'] as String),
      deadline_date: DateTime.parse(json['deadline_date'] as String),
      creation_date: DateTime.parse(json['creation_date'] as String),
      done_date: json.containsKey('done_date') ? DateTime.parse(json['done_date'] as String) : null,
    );
  }

  @override
  Map<String, dynamic> toJson() {
    return {
      'description': description,
      'location_id': location_id,
      'professional_amount': professional_amount,
      'laborer_amount': laborer_amount,
      'professional_minutes': professional_minutes,
      'laborer_minutes': laborer_minutes,
      'average_labor_cost': average_labor_cost,
      'done': done,
      'start_date': start_date.toIso8601String(),
      'deadline_date': deadline_date.toIso8601String(),
      'created_by': created_by,
      'creation_date': creation_date.toIso8601String(),
      'done_date': done_date?.toIso8601String(),
    };
  }
}

class Activity extends ActivityBase {
  final int id;

  const Activity({
    required this.id,
    required super.description,
    required super.location_id,
    required super.professional_amount,
    required super.laborer_amount,
    required super.professional_minutes,
    required super.laborer_minutes,
    required super.average_labor_cost,
    required super.done,
    required super.created_by,
    required super.creation_date,
    required super.start_date,
    required super.deadline_date,
    super.done_date,
  });

  factory Activity.fromJson(Map<String, dynamic> json) {
    return Activity(
      id: json['id'] as int,
      description: json['description'] as String,
      location_id: json['location_id'] as int,
      professional_amount: json['professional_amount'] as int,
      laborer_amount: json['laborer_amount'] as int,
      professional_minutes: json['professional_minutes'] as int,
      laborer_minutes: json['laborer_minutes'] as int,
      average_labor_cost: json['average_labor_cost'] as int,
      done: json['done'] as bool,
      created_by: json['created_by'] as int,
      start_date: DateTime.parse(json['start_date'] as String),
      deadline_date: DateTime.parse(json['deadline_date'] as String),
      creation_date: DateTime.parse(json['creation_date'] as String),
      done_date: json['done_date'] != null ? DateTime.parse(json['done_date'] as String) : null,
    );
  }

  @override
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'description': description,
      'location_id': location_id,
      'professional_amount': professional_amount,
      'laborer_amount': laborer_amount,
      'professional_minutes': professional_minutes,
      'laborer_minutes': laborer_minutes,
      'average_labor_cost': average_labor_cost,
      'start_date': start_date.toIso8601String(),
      'deadline_date': deadline_date.toIso8601String(),
      'done': done,
      'created_by': created_by,
      'creation_date': creation_date.toIso8601String(),
      'done_date': done_date?.toIso8601String(),
    };
  }
}

class ActivityClone implements JsonSerializable {
  final String description;
  final int professional_amount;
  final int laborer_amount;
  final int professional_minutes;
  final int laborer_minutes;
  final int average_labor_cost;
  final DateTime start_date;
  final DateTime deadline_date;
  final List<int> restriction;
  final List<int> materials;
  final List<double> materials_quantity;
  final bool same_project;

  const ActivityClone({
    required this.description,
    required this.professional_amount,
    required this.laborer_amount,
    required this.professional_minutes,
    required this.laborer_minutes,
    required this.average_labor_cost,
    required this.start_date,
    required this.deadline_date,
    required this.restriction,
    required this.materials,
    required this.materials_quantity,
    required this.same_project,
  });

  factory ActivityClone.fromJson(Map<String, dynamic> json) {
    return ActivityClone(
      description: json['description'] as String,
      professional_amount: json['professional_amount'] as int,
      laborer_amount: json['laborer_amount'] as int,
      professional_minutes: json['professional_minutes'] as int,
      laborer_minutes: json['laborer_minutes'] as int,
      average_labor_cost: json['average_labor_cost'] as int,
      start_date: DateTime.parse(json['start_date'] as String),
      deadline_date: DateTime.parse(json['deadline_date'] as String),
      restriction: (json['restriction'] as List<dynamic>).map((e) => e as int).toList(),
      materials: (json['materials'] as List<dynamic>).map((e) => e as int).toList(),
      materials_quantity: (json['materials_quantity'] as List<dynamic>).map((e) => (e as num).toDouble()).toList(),
      same_project: json['same_project'] as bool,
    );
  }

  @override
  Map<String, dynamic> toJson() {
    return {
      'description': description,
      'professional_amount': professional_amount,
      'laborer_amount': laborer_amount,
      'professional_minutes': professional_minutes,
      'laborer_minutes': laborer_minutes,
      'average_labor_cost': average_labor_cost,
      'start_date': start_date.toIso8601String(),
      'deadline_date': deadline_date.toIso8601String(),
      'restriction': restriction,
      'materials': materials,
      'materials_quantity': materials_quantity,
      'same_project': same_project,
    };
  }
}

File: Código-Fonte/fastapi/crud/activity.py
from sqlalchemy import and_, or_
from encryption.functions import authenticate_user
from crud.association.project_employee import get_projects_associated_with_logged_user
from crud.association.activity_restriction import get_activity_restrictions
from crud.association.activity_employee import get_activity_employees
from crud.association.project_activity import get_activities_by_project_id
from datetime import datetime, date
from schemas.activity import ActivityCreate, Activity, ActivityClone
from schemas.chunk import LocationChunkStatus, ChunkActivityStatus
import models as model

def get_all_activities(db: authenticate_user, skip:int=0, limit:int=1_500):
    return db.query(model.Activity).offset(skip).limit(limit).all()

def get_activity_by_id(db: authenticate_user, activity_id: int) -> Activity:
    return db.query(model.Activity).filter(model.Activity.id == activity_id).first()

def get_activity_by_creator_id(db: authenticate_user, user_id: int):
    return db.query(model.Activity).filter(model.Activity.created_by == user_id).all()

def get_activity_projects(db: authenticate_user, activity_id: int):
    _activity_projects = (
        db.query(model.ProjectActivityAssociation)
        .filter(
            model.ProjectActivityAssociation.activity_id == activity_id,
            model.ProjectActivityAssociation.deprecated == False
        )
        .all()
    )
    _projects = []
    for association in _activity_projects:
        _project = db.query(model.Project).filter(model.Project.id == association.project_id).first()
        if _project:
            _projects.append(_project)

    return _projects

def is_activity_related_with_logged_user(db: authenticate_user, user, activity_id: int) -> bool:
    user_id = user.get('id')

    if db.query(model.ActivityEmployeeAssociation).filter(
            model.ActivityEmployeeAssociation.activity_id == activity_id,
            model.ActivityEmployeeAssociation.employee_id == user_id
    ).first():
        return True

    activities = get_activities_with_restrictions_associated_with_logged_user(db=db, user=user)

    if any(i.id == activity_id for i in activities):
        return True

    for activity in activities:
        restrictions = get_activity_restrictions(db=db, activity_id=activity.id)
        if any(restricted_activity.id == activity_id for restricted_activity in restrictions):
            return True

    return False

def can_complete_activity(db: authenticate_user, activity_id: int) -> bool:
    restrictions = get_activity_restrictions(db=db, activity_id=activity_id)
    for restriction in restrictions:
        if restriction:
            if not restriction.done_date:
                return False

    _activity_chunks = (
        db.query(model.ActivityChunkAssociation)
        .filter(
            model.ActivityChunkAssociation.activity_id == activity_id,
            model.ActivityChunkAssociation.deprecated.is_(False),
        )
        .all()
    )

    for activity_chunk in _activity_chunks:
        _chunk: ChunkActivityStatus = db.query(model.ActivityChunkStatus).filter(
            model.ActivityChunkStatus.chunk_id == activity_chunk.chunk_id,
            model.ActivityChunkStatus.activity_id == activity_id
        ).first()
        if _chunk:
            if _chunk.status == False:
                return False

    return True

def is_user_responsible_for_activity_project(db: authenticate_user, activity_id: int, user_id: int) -> bool:
    activity_projects = (
        db.query(model.ProjectActivityAssociation)
        .filter(
            model.ProjectActivityAssociation.activity_id == activity_id,
            model.ProjectActivityAssociation.deprecated == False
        )
        .all()
    )

    for association in activity_projects:
        project = db.query(model.Project).filter(model.Project.id == association.project_id).first()
        if project and project.responsible_id == user_id:
            return True
            
    return False

def get_activities_associated_with_user_by_id(db: authenticate_user, user, user_id: int):
    query_employee_activities = db.query(model.Activity).join(
        model.ActivityEmployeeAssociation,
        model.ActivityEmployeeAssociation.activity_id == model.Activity.id
    ).filter(
        model.ActivityEmployeeAssociation.employee_id == user_id,
    )

    query_created_activities = db.query(model.Activity).filter(
        model.Activity.created_by == user_id,
    )

    query_project_employee_activities = db.query(model.Activity).join(
        model.ProjectActivityAssociation,
        model.ProjectActivityAssociation.activity_id == model.Activity.id
    ).join(
        model.ProjectEmployeeAssociation,
        model.ProjectEmployeeAssociation.project_id == model.ProjectActivityAssociation.project_id
    ).filter(
        model.ProjectEmployeeAssociation.employee_id == user_id,
    )

    query_project_rain_activities = db.query(model.Activity).join(
        model.ProjectRainAssociation,
        model.ProjectRainAssociation.rain_activity_id == model.Activity.id
    ).join(
        model.ProjectEmployeeAssociation,
        model.ProjectEmployeeAssociation.project_id == model.ProjectRainAssociation.project_id
    ).filter(
        model.ProjectEmployeeAssociation.employee_id == user_id,
    )

    activities = query_employee_activities.union(
        query_created_activities,
        query_project_employee_activities,
        query_project_rain_activities
    ).all()

    return activities

def get_activities_with_restrictions_associated_with_logged_user(db: authenticate_user, user):
    user_id = user.get('id')
    
    projects = get_projects_associated_with_logged_user(db=db, user=user)
    activities = set()

    # Para cada projeto, recupera suas atividades
    for project in projects:
        project_activities = get_activities_by_project_id(db=db, project_id=project.id)
        for activity in project_activities:
            activities.add(activity)
            
            restrictions = get_activity_restrictions(db=db, activity_id=activity.id)
            for restriction in restrictions:
                activities.add(restriction)

    clean_activities = list(activities)

    return clean_activities

# created this but not actually used anywere ¯\_(ツ)_/¯
def mark_activity_as_done(db: authenticate_user, activity_id: int):
    activity = get_activity_by_id(db=db, activity_id=activity_id)
    if activity is None:
        return None 
    activity.done = True
    db.commit()
    db.refresh(activity)
    return activity

def update_activity(db: authenticate_user, activity_id: int, updated_activity: ActivityCreate):
    activity = get_activity_by_id(db=db, activity_id=activity_id)
    if activity is None:
        return None

    activity.description = updated_activity.description
    activity.location_id = updated_activity.location_id
    activity.professional_amount = updated_activity.professional_amount
    activity.laborer_amount = updated_activity.laborer_amount
    activity.professional_minutes = updated_activity.professional_minutes
    activity.laborer_minutes = updated_activity.laborer_minutes
    activity.average_labor_cost = updated_activity.average_labor_cost
    
    db.commit()
    db.refresh(activity)
    return activity

def create_activity(db: authenticate_user, user, activity: ActivityCreate):
    user_id = user.get('id')
    _activity = model.Activity(
        description = activity.description,
        location_id = activity.location_id,
        professional_amount = activity.professional_amount,
        laborer_amount = activity.laborer_amount,
        professional_minutes = activity.professional_minutes,
        average_labor_cost = activity.average_labor_cost,
        laborer_minutes = activity.laborer_minutes,
        created_by = user_id,
        start_date=activity.start_date,
        deadline_date=activity.deadline_date,
        creation_date = datetime.now(),
        done = False
    )
    db.add(_activity)
    db.commit()
    db.refresh(_activity)
    return _activity

def create_clone_activity(db: authenticate_user, user, activity: ActivityClone, location_id: int):
    user_id = user.get('id')
    _activity = model.Activity(
        description = activity.description,
        location_id = location_id,
        professional_amount = activity.professional_amount,
        laborer_amount = activity.laborer_amount,
        professional_minutes = activity.professional_minutes,
        laborer_minutes = activity.laborer_minutes,
        average_labor_cost = activity.average_labor_cost,
        created_by = user_id,
        start_date=activity.start_date,
        deadline_date=activity.deadline_date,
        creation_date = datetime.now(),
        done = False
    )
    db.add(_activity)
    db.commit()
    db.refresh(_activity)
    return _activity


def update_activity_property(db: authenticate_user, activity_id: int, update_data: dict):
    activity: Activity = get_activity_by_id(db=db, activity_id=activity_id)

    for i in update_data:
        if str(i) in ["id", "creation_date", "created_by", "done_date"]:
            return None

    if "done" in update_data:
        if not can_complete_activity(db=db, activity_id=activity_id):
            return None
        else:
            setattr(activity, "done_date", date.today())
            setattr(activity, "done", True)
            update_data.pop("done", None)

    for field, value in update_data.items():
        if hasattr(activity, field):
            setattr(activity, field, value)
        else:
            raise ValueError(f"Field '{field}' does not exist on the Activity model")

    db.commit()
    db.refresh(activity)
    return activity if activity else None